File: behavioral/typescript/memento_structured/game_state.ts
export class GameStateMemento {
  private level: number;
  private health: number;
  private score: number;
  private position: string;
  
  constructor(level: number, health: number, score: number, position: string) {
    this.level = level;
    this.health = health;
    this.score = score;
    this.position = position;
  }
  
  getLevel(): number { return this.level; }
  getHealth(): number { return this.health; }
  getScore(): number { return this.score; }
  getPosition(): string { return this.position; }
}

export class GameCharacter {
  private level: number = 1;
  private health: number = 100;
  private score: number = 0;
  private position: string = 'Start';
  
  moveTo(newPosition: string): void {
    this.position = newPosition;
    console.log(`Moved to ${this.position}`);
  }
  
  takeDamage(damage: number): void {
    this.health -= damage;
    if (this.health < 0) this.health = 0;
    console.log(`Took ${damage} damage. Health: ${this.health}`);
  }
  
  gainScore(points: number): void {
    this.score += points;
    console.log(`Gained ${points} points. Score: ${this.score}`);
  }
  
  levelUp(): void {
    this.level++;
    this.health = 100;
    console.log(`Level up! Now at level ${this.level}`);
  }
  
  displayStatus(): void {
    console.log('=== Character Status ===');
    console.log(`Level: ${this.level}`);
    console.log(`Health: ${this.health}`);
    console.log(`Score: ${this.score}`);
    console.log(`Position: ${this.position}`);
    console.log('=====================');
  }
  
  saveState(): GameStateMemento {
    console.log('Saving game state...');
    return new GameStateMemento(this.level, this.health, this.score, this.position);
  }
  
  restoreState(memento: GameStateMemento): void {
    this.level = memento.getLevel();
    this.health = memento.getHealth();
    this.score = memento.getScore();
    this.position = memento.getPosition();
    console.log('Game state restored!');
    this.displayStatus();
  }
}

export class GameSaveManager {
  private saves: GameStateMemento[] = [];
  
  createSave(save: GameStateMemento): void {
    this.saves.push(save);
    console.log(`Save created. Total saves: ${this.saves.length}`);
  }
  
  loadSave(index: number): GameStateMemento | null {
    if (index >= 0 && index < this.saves.length) {
      console.log(`Loading save ${index + 1}...`);
      return this.saves[index];
    }
    return null;
  }
  
  getSaveCount(): number {
    return this.saves.length;
  }
}

File: behavioral/dart/memento_structured/game_state.dart
class GameStateMemento {
  final int level;
  final int health;
  final int score;
  final String position;
  
  GameStateMemento(this.level, this.health, this.score, this.position);
  
  int get getLevel => level;
  int get getHealth => health;
  int get getScore => score;
  String get getPosition => position;
}

class GameCharacter {
  int _level = 1;
  int _health = 100;
  int _score = 0;
  String _position = 'Start';
  
  void moveTo(String newPosition) {
    _position = newPosition;
    print('Moved to $_position');
  }
  
  void takeDamage(int damage) {
    _health -= damage;
    if (_health < 0) _health = 0;
    print('Took $damage damage. Health: $_health');
  }
  
  void gainScore(int points) {
    _score += points;
    print('Gained $points points. Score: $_score');
  }
  
  void levelUp() {
    _level++;
    _health = 100;
    print('Level up! Now at level $_level');
  }
  
  void displayStatus() {
    print('=== Character Status ===');
    print('Level: $_level');
    print('Health: $_health');
    print('Score: $_score');
    print('Position: $_position');
    print('=====================');
  }
  
  GameStateMemento saveState() {
    print('Saving game state...');
    return GameStateMemento(_level, _health, _score, _position);
  }
  
  void restoreState(GameStateMemento memento) {
    _level = memento.getLevel;
    _health = memento.getHealth;
    _score = memento.getScore;
    _position = memento.getPosition;
    print('Game state restored!');
    displayStatus();
  }
}

class GameSaveManager {
  final List<GameStateMemento> _saves = [];
  
  void createSave(GameStateMemento save) {
    _saves.add(save);
    print('Save created. Total saves: ${_saves.length}');
  }
  
  GameStateMemento? loadSave(int index) {
    if (index >= 0 && index < _saves.length) {
      print('Loading save ${index + 1}...');
      return _saves[index];
    }
    return null;
  }
  
  int get saveCount => _saves.length;
}
